[Code coverage] Improve the http module coverage.

knightburton opened the pull request:
- Add new test cases to the existing files.
- Add new test file to cover the request response messages.

Note: The setTimeout function from the outgoungMessage is deleted,
because every parents are implements their own setTimeout function.

IoT.js-DCO-1.0-Signed-off-by: Imre Kiss kissi.szeged@partner.samsung.com

yichoi commented:
@knightburton Please rebase it to the master

daeyeon commented on test/run_pass/test_net_httpserver_timeout.js:
```diff
@@ -25,6 +25,8 @@ var server = http.createServer(function(req, res) {
 
 server.listen(3003);
+// Overwrite the default timeout value.
+server.setTimeout(100);
```
There is no verification for this test.

knightburton replied:
I've added an assert below this line.

daeyeon replied:
I suggest the test like below.

```
var timer = setTimeout(function () {
    defaultTimeout = true;
}, server.timeout);

server.setTimeout(XXX, function(socket) {
    ....
    assert.equal(defaultTimeout, false);
    clearTimeout(timer);
```

daeyeon commented on src/js/http_outgoing.js:
```diff
@@ -152,8 +152,8 @@ OutgoingMessage.prototype._storeHeader = function(statusLine) {
-  if (this._headers === null) {
-    this._headers = {};
+  if (util.isNull(this._headers)) {
+    this._headers = [];
```
First of all, this commit is to improve code coverage. So please don't fix any original code. Plus, I don't agree this change. Kindly let me know why using `Array` is needed. Seeing the code below, `this._headers` is still used as `Object` although you set it as `Array`.

knightburton replied:
You are right! I will separate the coverage and fix in the future. The other hand, you are right again, there is no point to use array instead of object, this was my inattention.

daeyeon replied:
#798 is related to this discussion.

yichoi commented:
@knightburton please rebase it to the master

knightburton commented:
@daeyeon I've updated the PR. I would like to know what do you think about it. First of all, I've removed the code fixes (I'll update them in a separate PR later), and isolated every test case into a server-request pair in the request-response test file. In the other hand, I removed the server default timeout overwrite test for now, because I think neither of our solution is not the best. I would like to update this patch with a proper timeout test and invalid utf-8 test later.

daeyeon commented on test/run_pass/test_net_http_request_response.js:
```diff
@@ -0,0 +1,206 @@
+  });
+
+  response.on('end', function() {
+    assert.equal(str, 'default');
```
This seems too natural connection failure case to test. Do we need to test if our api doesn't work when user doesn't fill up the mandatory field for the api?

knightburton replied:
Yes, I think it is necessary. I think we should test the default, natural cases. too.

daeyeon replied:
In the case, it seems right to check if the api doesn't crash, or else if it throws any information for the wrong usage. However, assert here checks str although it's even never be called. Confusing.

daeyeon replied:
You maybe misunderstood `options`. It's mandatory, so that the api doesn't need to gurantee the case where user doesn't fill it up. If the given `options` is just an object having nothing, the api fill it up with default values. https://github.com/Samsung/iotjs/blob/master/docs/api/IoT.js-API-HTTP.md#httpgetoptions-callback

zherczeg replied:
Should it throw an error if no options passed then?

knightburton replied:
No, it should not. It is just fills up the options values with default parameters. That is what i've tested, if we can't define any options value, the connection should be fine.

daeyeon replied:
Throwing an error is needed. e.g) calling `server0.listen()` without the first mandatory param, port, throws an error. The criteria is the API reference, not the current observation.

knightburton replied:
Sorry for the late response, then what should I do with this test? Should I remove it for now?

daeyeon replied:
Yes, please remove this test for now. It'd be better to treat it on another commit for changing the condition of `options` to optional.

knightburton replied:
Done, removed the test.

daeyeon commented on test/run_pass/test_net_http_request_response.js:
```diff
@@ -0,0 +1,206 @@
+  path: '/'
+}, function(response) {
+  response.on('end', function() {
+    assert.equal(response.statusCode, 200);
```
Please clearly use `assert` on what you check for each test. Since this compares the status code, it seems to assume that the test needs any data from server. I suggest the below pattern. The same problem is found elsewhere in this code.

```
var isTestCase1Called = false;
var request1 = http.get(function(res) {
  ....
  res.on('end', function () {
      isTestCase1Called = true;
  });
  ....
});


process.on('exit', function () {
  assert.equal(isTestCase1Called, false);
});
``

knightburton replied:
I understand what are you suggest, but also I think this test case is fine because the communication should be OK without any body data and that is why there is the status code check.

zherczeg commented on test/run_pass/test_net_headers.js:
```diff
@@ -29,6 +29,7 @@ var server = http.createServer(function (req, res) {
   });
 
   var endHandler = function () {
+    res.removeHeader('h1'); // this should return.
```
Put comment before this line.

daeyeon commented:
LGTM
